src/hotspot/share/runtime/crac.cpp

```diff
@@ -488,10 +488,16 @@ void crac::restore() {
   }
 
   if (_crengine) {
-    _crengine_args[1] = "restore";
-    add_crengine_arg(CRaCRestoreFrom);
-    os::execv(_crengine, _crengine_args);
-    warning("cannot execute \"%s restore ...\" (%s)", _crengine, os::strerror(errno));
+    if (is_dynamic_library(_crengine)) {
+      // This function should not return on success
+      call_crengine_library(false, CRaCRestoreFrom);
+      warning("cannot restore from dynamic library %s", _crengine);
+    } else {
+      _crengine_args[1] = "restore";
+      add_crengine_arg(CRaCRestoreFrom);
+      os::execv(_crengine, _crengine_args);
+      warning("cannot execute \"%s restore ...\" (%s)", _crengine, os::strerror(errno));
+    }
```

Comment:
It looks like we can have a single interface with dynamic libs. The lib may always spawn a process if required.

And to simplify the implementation even more, we can open the lib earlier that the checkpoint.

Reply:
> It looks like we can have a single interface with dynamic libs. The lib may always spawn a process if required.

Are you suggesting that we should turn criuengine/pauseengine/simengine to dynamic libraries and use only that?

> And to simplify the implementation even more, we can open the lib earlier that the checkpoint.

In other words, open it and keep open = leak it?

Reply:
> From the implementation, it will be enough to dlopen the engine on the VM start.

OK, in fact it is probably better to fail early.